relay-general/src/protocol/measurements.rs

```diff
@@ -0,0 +1,199 @@
+                let mut measurements = Object::<Measurement>::new();
+
+                for (raw_name, raw_observed_value) in items.into_iter() {
+                    let observed_value: Annotated<f64> = match raw_observed_value {
+                        Annotated(Some(Value::Object(bag)), mut object_meta) => {
+                            match bag.get("value") {
+                                Some(Annotated(Some(Value::I64(value)), meta)) => {
+                                    Annotated(Some(*value as f64), meta.clone())
+                                }
+                                Some(Annotated(Some(Value::U64(value)), meta)) => {
+                                    Annotated(Some(*value as f64), meta.clone())
+                                }
+                                Some(Annotated(Some(Value::F64(value)), meta)) => {
+                                    Annotated(Some(*value), meta.clone())
+                                }
+                                Some(Annotated(value, meta)) => {
+                                    measurement_meta.add_error(Error::expected(
+                                        format!(
+                                            "measurement value for '{}' to be a number",
+                                            raw_name
+                                        )
+                                        .as_str(),
+                                    ));
+
+                                    let mut meta = meta.clone();
+                                    meta.add_error(Error::expected("number"));
+                                    meta.set_original_value(value.clone());
+                                    Annotated(None, meta)
+                                }
+                                None => {
+                                    object_meta.add_error(Error::expected("value"));
+                                    Annotated(None, object_meta)
+                                }
+                            }
+                        }
+                        Annotated(value, meta) => {
+                            let mut meta = meta.clone();
+                            meta.add_error(Error::expected("object"));
+                            meta.set_original_value(value);
+                            Annotated(None, meta.clone())
+                        }
+                    };
+
+                    let measurement_name = raw_name.trim().to_lowercase();
+
+                    if MEASUREMENT_NAME.is_match(&measurement_name) {
+                        if observed_value.value().is_some() {
+                            let measurement = Measurement {
+                                value: observed_value,
+                            };
+
+                            measurements.insert(measurement_name, Annotated::new(measurement));
+                        }
+                    } else {
+                        measurement_meta.add_error(Error::expected(
```

Comment:
For such errors, consider using `Error::invalid(format!(...))`. Note how it takes the string directly.

Reply:
Done.